File: domain/src/main/java/vn/kingbee/domain/core/exception/RetrofitException.java
package vn.kingbee.domain.core.exception;

import java.io.IOException;

import retrofit2.HttpException;
import retrofit2.Response;

public class RetrofitException extends RuntimeException {
    private String url;
    private Response response;
    private RetrofitException.Kind kind;

    RetrofitException(String message, String url, Response response, RetrofitException.Kind kind, Throwable exception) {
        super(message, exception);
        this.url = url;
        this.response = response;
        this.kind = kind;
    }

    public static RetrofitException httpError(String url, Response response) {
        String message = response.code() + " " + response.message();
        return new RetrofitException(message, url, response, RetrofitException.Kind.HTTP, null);
    }

    public static RetrofitException networkError(String url, Response response, IOException exception) {
        return new RetrofitException(exception.getMessage(), url, response, RetrofitException.Kind.NETWORK, exception);
    }

    public static RetrofitException unexpectedError(String url, Response response, Throwable exception) {
        return new RetrofitException(exception.getMessage(), url, response, RetrofitException.Kind.UNEXPECTED, exception);
    }

    public static RetrofitException fromThrowable(Throwable throwable) {
        if (throwable instanceof HttpException) {
            HttpException httpException = (HttpException) throwable;
            Response response = httpException.response();
            return httpError(response.raw().request().url().toString(), response);
        } else {
            return throwable instanceof IOException ? networkError(null, null, (IOException) throwable) : unexpectedError(null, null, throwable);
        }
    }

    public String getUrl() {
        return this.url;
    }

    public Response getResponse() {
        return this.response;
    }

    public RetrofitException.Kind getKind() {
        return this.kind;
    }

    public enum Kind {
        NETWORK,
        HTTP,
        UNEXPECTED;

        Kind() {
        }
    }
}

File: domain/src/main/java/vn/kingbee/domain/core/exception/RetrofitException.kt
package vn.kingbee.domain.core.exception

import java.io.IOException

import retrofit2.HttpException
import retrofit2.Response

class RetrofitException internal constructor(
    message: String?,
    val url: String?,
    val response: Response<*>?,
    val kind: Kind,
    exception: Throwable?
) : RuntimeException(message, exception) {

    enum class Kind {
        NETWORK,
        HTTP,
        UNEXPECTED
    }

    companion object {

        fun httpError(url: String, response: Response<*>): RetrofitException {
            val message = response.code().toString() + " " + response.message()
            return RetrofitException(message, url, response, Kind.HTTP, null)
        }

        fun networkError(url: String?, response: Response<*>?, exception: IOException): RetrofitException {
            return RetrofitException(exception.message, url, response, Kind.NETWORK, exception)
        }

        fun unexpectedError(url: String?, response: Response<*>?, exception: Throwable): RetrofitException {
            return RetrofitException(exception.message, url, response, Kind.UNEXPECTED, exception)
        }

        fun fromThrowable(throwable: Throwable): RetrofitException {
            if (throwable is HttpException) {
                val response = throwable.response()
                return httpError(response.raw().request().url().toString(), response)
            } else {
                return if (throwable is IOException) networkError(null, null, throwable) else unexpectedError(
                    null,
                    null,
                    throwable
                )
            }
        }
    }
}
